



library(Seurat)
library(hypeR)
library(screp)
library(stringr)

######set up annotation files
SYMBOLS<-unlist(as.list(org.Mm.egSYMBOL))
Entrez<-unlist(as.list(org.Mm.egENSEMBL))
x<-names(Entrez)
names(x)<-Entrez
Entrez<-x
Ensembl<-unlist(as.list(org.Mm.egENSEMBL))
mart <- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))
x<-rownames(young.combined@assays$RNA@counts)
x<-unique(c(rownames(young.combined@assays$RNA@counts),rownames(mm126191@assays$RNA@counts)))
x<-strsplit(x,"-")
x<-unlist(x)
x<-x[seq(1,length(x),by=2)]
x<- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","mgi_symbol"),values=x,mart= mart)
y<-x[,2]
names(y)<-x[,1]
MGI<-y
MGI.rev<-names(MGI)
names(MGI.rev)<-MGI

x <- Term(GOTERM)
goterms<-names(x)
names(goterms)<-x


s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
s.genes<-str_to_title(s.genes)
g2m.genes<-str_to_title(g2m.genes)

#####add descriptors to metadata from 
mm126191@meta.data$age<-"old"
mm126191@meta.data$position<-mm126191@meta.data$sample
#######get counts including introns 
z<-read.csv("126191_out_incl_introns_genematrix.csv",as.is=T,row.names=1)
x<-read.csv("126191_out_genematrix.csv",as.is=T,row.names=1)
z1<-read.csv("126254_out_incl_introns_genematrix.csv",as.is=T,row.names=1)
x1<-read.csv("126254_out_genematrix.csv",as.is=T,row.names=1)
z2<-read.csv("126275_out_incl_introns_genematrix.csv",as.is=T,row.names=1)
x2<-read.csv("126275_out_genematrix.csv",as.is=T,row.names=1)

## calculate out the ratio of  exon+intron to intron counts and remove infinte/nan cases
m<-z-x
m1<-z1-x1
m2<-z2-x2

test<-apply(z,1,sum)/apply(m,1,sum)
test<-test[-(which(is.infinite(test)))]
test<-test[-(which(is.nan(test)))]

test1<-apply(z1,1,sum)/apply(m1,1,sum)
test1<-test1[-(which(is.infinite(test1)))]
test1<-test1[-(which(is.nan(test1)))]

test2<-apply(z2,1,sum)/apply(m2,1,sum)
test2<-test2[-(which(is.infinite(test2)))]
test2<-test2[-(which(is.nan(test2)))]

###remove those genes with a ratio greater than 100
### We reason that mRNA transcibed by the cell should have a fairly high level of intronic sequences vs non-nuclear mRNA. 
### By taking this approach we can weed out the greater portion of the ambient RNA

moo<-names(test[which(test>100)])
moo1<-names(test1[which(test1>100)])
moo2<-names(test2[which(test2>100)])

arna<-unique(c(moo,moo1,moo2))
genes<-setdiff(unique(c(names(test),names(test1),names(test2))),arna)

# create seurat objects for each age. Get metadata from object generated by Farhad. 
##old(18months)
y<-mm126191@meta.data
m<-rownames(mm126191@meta.data)
m<-unlist(str_split(m,"_"))
m<-m[seq(2,length(m),2)]
rownames(y)<-m
y<-y[,10:11]
z<-z[,rownames(y)]
z<-z[genes,]
m<-rownames(z)
x<-strsplit(m,"_")
x<-unlist(x)
x<-x[seq(2,length(x),by=2)]
m<-duplicated(x)
z<-z[which(!m),]
rownames(z)<-x[which(!m)]

##young( 3months)
y0<-young.combined@meta.data
m<-rownames(young.combined@meta.data)
m<-unlist(str_split(m,"_"))
m<-m[seq(2,length(m),2)]
rownames(y0)<-m
z3<-cbind(z1,z2)
z3<-z3[,rownames(y0)]
z3<-z3[genes,]
m<-rownames(z3)
x<-strsplit(m,"_")
x<-unlist(x)
x<-x[seq(2,length(x),by=2)]
m<-duplicated(x)
z3<-z3[which(!m),]
rownames(z3)<-x[which(!m)]
y0<-y0[,9:10]

y0<-CreateSeuratObject(z3,meta.data=y0,min.cells=5,min.features=1000)
ol0<-CreateSeuratObject(z,meta.data=y,min.cells=5,min.features=1000)

## add cell cyle info to objects
y0<-CellCycleScoring(y0, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
y0<-PercentageFeatureSet(y0, "^mt-", col.name = "percent_mito")
ol0<-CellCycleScoring(ol0, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
ol0<-PercentageFeatureSet(ol0, "^mt-", col.name = "percent_mito")


###analysis and cell annotation
#####young 
y1<-SCTransform(y1,vars.to.regress=c("S.Score", "G2M.Score"))
y1<-RunPCA(y1)
ElbowPlot(y1)
y1<-RunUMAP(y1,dims=1:10)
y1<-FindNeighbors(y1,dims=1:10)
y1<-FindClusters(y1,resolution = 0.8)
DimPlot(y1,label=T)
DimPlot(y1,split.by="position",label=T)

VlnPlot(y1,c("Ptprc","Pecam1"))

# old

o1<-SCTransform(o1,vars.to.regress=c("S.Score", "G2M.Score"))
o1<-RunPCA(o1)
ElbowPlot(o1)
o1<-RunUMAP(o1,dims=1:10)
o1<-FindNeighbors(o1,dims=1:10)
o1<-FindClusters(o1,resolution = 0.8)
DimPlot(o1,label=T)
DimPlot(o1,group.by="position")


o1@misc$markers<-FindAllMarkers(o1,only.pos=T)
y1@misc$markers<-FindAllMarkers(y1,only.pos=T)

o1@misc$markers<-o1@misc$markers[o1@misc$markers$p_val_adj<0.1,]
y1@misc$markers<-y1@misc$markers[y1@misc$markers$p_val_adj<0.1,]

########integrate ages
z<-list(o1,y1)
z1<-SelectIntegrationFeatures(object.list=z,nfeatures=3000)
z<-PrepSCTIntegration(object.list=z,anchor.features=z1)
za<-FindIntegrationAnchors(object.list=z,normalization.method = "SCT",
                           anchor.features = z1,reduction="cca",k.anchor=30)
obj<-IntegrateData(anchorset=za,normalization.method = "SCT")

obj<-RunPCA(obj)
ElbowPlot(obj)
obj<-RunUMAP(obj,dims=1:5)
obj<-FindNeighbors(obj,dims=1:5)
obj<-FindClusters(obj,resolution = 0.5)
DefaultAssay(obj)<-"SCT"
obj<-PrepSCTFindMarkers(obj)
obj@misc$markers<-FindAllMarkers(obj,only.pos=T)
obj@misc$markers<-obj@misc$markers[obj@misc$markers$p_val_adj<0.1,]
obj@misc$markers<-FindAllMarkers(obj,only.pos=T)


###position is a main contributor to variance makes identifying cells a bit hard, decided to regress 
###position out, use a more conservative filter for library size, identify cells and then map onto the object without position regression 
### Reference datasets for cell annotation not based on single nuclear RNA-seq, so sought genes to help identify cell types
###genes used for annotation of cell types
##Negr1, Dcn, Igfbp2  higher in Fibroblasts over endothelial
## Pdzd2, Myh11 higher in endo/SMC over fibro
##markers PMID: 33343397
##pericytes ebf1 PMID:34272603; Rpbj PMID: 31249304
###Fibroblasts
## PMID: 25013174; PMID: 32403949
#total of 13 endothelial markers and 
y2<-subset(y0,subset=nCount_RNA>30000)
o2<-subset(ol0,subset=nCount_RNA>30000)

##young
y2<-SCTransform(y2,vars.to.regress=c("S.Score", "G2M.Score","position"))
y2<-RunPCA(y2)
ElbowPlot(y2)
y2<-RunUMAP(y2,dims=1:10)
y2<-FindNeighbors(y2,dims=1:10)
y2<-FindClusters(y2,resolution = 0.8)
DimPlot(y2,label=T)

x<-DotPlot(y2,features=c("Negr1","Igfbp2","Ppp1r1a","Serpine2",
                          "Runx1t1","Epha5","Tjp1","Nfix","Meis2",
                          "Col1a2","Dcn","Ebf1","Pdzd2",
                          "Nrp1","Rbpj","Angpt1","Runx1","Fbln5","Vegfc",
                          "Myh11","Pecam1","Abcc9","Ets1","Erg"),
           group.by="celltype")

x+coord_flip()
## cell type is main factor in driving the cells apart in UMAP now, however position still pretty strong driver.

# old

o2<-SCTransform(o2,vars.to.regress=c("S.Score", "G2M.Score","position"))
o2<-RunPCA(o2)
ElbowPlot(o2)
o2<-RunUMAP(o2,dims=1:20)
o2<-FindNeighbors(o2,dims=1:10)
o2<-FindClusters(o2,resolution = 0.8)
DimPlot(o2,label=T)

x<-DotPlot(o2,features=c("Negr1","Igfbp2","Ppp1r1a","Serpine2",
                          "Runx1t1","Epha5","Tjp1","Nfix","Meis2",
                          "Col1a2","Dcn","Ebf1","Pdzd2",
                          "Nrp1","Rbpj","Angpt1","Runx1","Fbln5","Vegfc",
                          "Myh11","Pecam1","Abcc9","Ets1","Erg"),
           group.by="celltype")
x+coord_flip()

## Old cells are more muddled than young cells not clean seperation like in young
## hopefully the young cells will act as anchors during the integration
########integrate ages
z<-list(o2,y2)
z1<-SelectIntegrationFeatures(object.list=z,nfeatures=3000)
z<-PrepSCTIntegration(object.list=z,anchor.features=z1)
za<-FindIntegrationAnchors(object.list=z,normalization.method = "SCT",
                           anchor.features = z1,reduction="cca",k.anchor=30)
obj2<-IntegrateData(anchorset=za,normalization.method = "SCT")

obj2<-RunPCA(obj2)
ElbowPlot(obj2)
obj2<-RunUMAP(obj2,dims=1:10)
obj2<-FindNeighbors(obj2,dims=1:10)
obj2<-FindClusters(obj2,resolution = 0.5)
DimPlot(obj2,label=T)

x<-DotPlot(obj2,features=c("Negr1","Igfbp2","Ppp1r1a","Serpine2",
                          "Runx1t1","Epha5","Tjp1","Nfix","Meis2",
                          "Col1a2","Dcn","Ebf1","Pdzd2",
                          "Nrp1","Rbpj","Angpt1","Runx1","Fbln5","Vegfc",
                          "Myh11","Pecam1","Abcc9","Ets1","Erg"),
           group.by="celltype")
x+coord_flip()
## success, cell ttypes fell out cleanly, adding to metadata 
z<-as.vector(y2$seurat_clusters)
z<-replace(z,which(z=="0"),"Fibro1")
z<-replace(z,which(z=="1"),"Endo1")
z<-replace(z,which(z=="2"),"Endo2")
z<-replace(z,which(z=="3"),"Fibro2")
z<-replace(z,which(z=="4"),"Endo3")
z<-replace(z,which(z=="5"),"Fibro3")
names(z)<-rownames(y2@meta.data)
y2$celltype<-z
names(z)<-rownames(y2@meta.data)
y1$celltype<-z

z<-as.vector(o2$seurat_clusters)
z<-replace(z,which(z=="0"),"Fibro1")
z<-replace(z,which(z=="1"),"Endo1")
z<-replace(z,which(z=="2"),"Endo2")
z<-replace(z,which(z=="3"),"Fibro2")
z<-replace(z,which(z=="4"),"Endo3")
z<-replace(z,which(z=="5"),"Fibro3")
o2$celltype<-z
names(z)<-rownames(o2@meta.data)
o1$celltype<-z

z<-as.vector(obj2$seurat_clusters)
z<-replace(z,which(z=="0"),"Fibro1")
z<-replace(z,which(z=="1"),"Endo1")
z<-replace(z,which(z=="2"),"Endo2")
z<-replace(z,which(z=="3"),"Fibro2")
z<-replace(z,which(z=="4"),"Fibro3")
z<-replace(z,which(z=="5"),"Fibro4")
z<-replace(z,which(z=="6"),"Endo3")
z<-replace(z,which(z=="7"),"Fibro5")
obj2$celltype<-z
names(z)<-rownames(obj2@meta.data)
obj$celltype<-z

DimPlot(obj,group.by="celltype",label=T,split.by="position")
DimPlot(obj,group.by="celltype",label=T)
DimPlot(obj,group.by="position")
DimPlot(obj,group.by="age")
DimPlot(obj,label=T)
obj<-SetIdent(obj,value="celltype")
obj@misc$celltype<-FindAllMarkers(obj,only.pos=T)

###OK cell types now assigned to each object and looks good 

x<-DotPlot(obj,features=c("Negr1","Igfbp2","Ppp1r1a","Serpine2",
                          "Runx1t1","Epha5","Tjp1","Nfix","Meis2",
                          "Col1a2","Dcn","Ebf1","Pdzd2",
                          "Nrp1","Rbpj","Angpt1","Runx1","Fbln5","Vegfc",
                          "Myh11","Pecam1","Abcc9","Ets1","Erg"),
           group.by="celltype")
x+coord_flip()

#####enrichment testing
GOBP <- msigdb_gsets(species="Mus musculus", category="C5", subcategory="BP")
eres<-enrich_test(markers_df=obj@misc$markers,species_x="Mus musculus",genome_genes=42937)
go_vis_sc<-GO_visualization(eres$Enriched_df,markers_df=obj@misc$markers,GOcats=GOBP,goterms=goterms,numcats=10,org_db="org.Mm.eg.db")
go_vis_sc$plot
data(RP)
RP<-RPprep(RP,"Mus")
data(RPR)
RP_ready<-reactome_prep(eres$Enriched_df,RP=RP,RP_adj=RPR)
react_vis<-reactome_visualization(RP_ready,RPR,RP=RP)
react_vis$plot_object

write.csv(go_vis_sc$GO_sem,"GO_table.csv")
write.csv(RP_ready,"Reactome_table.csv")

eres_y<-enrich_test(markers_df=y1@misc$markers,species_x="Mus musculus",genome_genes=42937)
go_vis_y<-GO_visualization(eres_y$Enriched_df,markers_df=y1@misc$markers,GOcats=GOBP,goterms=goterms,numcats=10,org_db="org.Mm.eg.db")
go_vis_y$plot


eres_o<-enrich_test(markers_df=o1@misc$markers,species_x="Mus musculus",genome_genes=42937)
go_vis_o<-GO_visualization(eres_o$Enriched_df,markers_df=o1@misc$markers,GOcats=GOBP,goterms=goterms,numcats=10,org_db="org.Mm.eg.db")
go_vis_o$plot

write.csv(go_vis_y$GO_sem,"GO_y_table.csv")
write.csv(go_vis_o$GO_sem,"GO_o_table.csv")










